File: android/app/src/main/java/com/gameshelf/data/secrets/AppSecrets.kt
package com.gameshelf.data.secrets

import android.content.Context
import com.gameshelf.BuildConfig
import com.gameshelf.R
import com.gameshelf.data.net.UserFacingError

/**
 * Las claves que vienen del archivo de configuracion, si lo hay.
 *
 * Entran por `Config/Secrets.xcconfig` (el mismo archivo que usa la version de
 * iOS, ignorado por git), el script de Gradle las mete en `BuildConfig` y se
 * leen aca en tiempo de ejecucion. Es el equivalente exacto del rodeo de iOS
 * (xcconfig -> Info.plist -> Bundle).
 *
 * **Ya no es la fuente principal.** Desde que cada cuenta pide lo suyo al
 * conectarse, esto es solo el respaldo: lo usa
 * [com.gameshelf.data.secrets.SteamCredentialsStore] cuando no hay nada
 * guardado, para que quien ya tuviera el archivo siga funcionando sin tener
 * que volver a conectar nada. Lo que el usuario conecta desde la app siempre
 * gana.
 *
 * Se mantiene la propiedad de siempre: **si falta una clave el proyecto
 * compila igual**. Lo que cambio es que ya no hace falta que ninguna este.
 *
 * Ver `docs/decisiones/002-gestion-de-secretos.md`.
 */
object AppSecrets {

  /**
   * Claves disponibles, con el nombre exacto que tienen en el archivo de
   * configuracion.
   */
  enum class Key(val configName: String) {
    STEAM_API_KEY("STEAM_API_KEY"),
    STEAM_ID("STEAM_ID"),
    ITAD_API_KEY("ITAD_API_KEY"),
  }

  /**
   * Error con un mensaje accionable, para mostrarlo en pantalla en vez de
   * dejar que la app se caiga.
   */
  data class MissingSecretError(val key: Key) : Exception(), UserFacingError {
    override fun message(context: Context): String =
      context.getString(R.string.secret_missing, key.configName)

    override fun recovery(context: Context): String =
      context.getString(R.string.secret_missing_recovery, key.configName)

    override val message: String get() = "Falta la clave ${key.configName}"
  }

  /**
   * De donde salen los valores.
   *
   * Se deja inyectar para las pruebas: en produccion es `BuildConfig`, pero un
   * test no puede recompilar el proyecto para cambiarlo.
   */
  fun interface Source {
    fun value(key: Key): String?
  }

  /** La fuente real: lo que Gradle escribio en `BuildConfig`. */
  val buildConfigSource = Source { key ->
    when (key) {
      Key.STEAM_API_KEY -> BuildConfig.STEAM_API_KEY
      Key.STEAM_ID -> BuildConfig.STEAM_ID
      Key.ITAD_API_KEY -> BuildConfig.ITAD_API_KEY
    }
  }

  /**
   * Devuelve el valor de una clave.
   *
   * @throws MissingSecretError si no esta o esta vacia.
   */
  fun value(key: Key, source: Source = buildConfigSource): String {
    val bruto = source.value(key)?.trim()
    if (bruto.isNullOrEmpty()) throw MissingSecretError(key)
    return bruto
  }
}

File: GameShelf/Services/AppSecrets.swift
//
//  AppSecrets.swift
//  GameShelf
//

import Foundation

/// Acceso a las claves de API que la app necesita.
///
/// Los valores no viven en el codigo: entran por `Config/Secrets.xcconfig`
/// (ignorado por git), pasan al Info.plist como variables de build y se leen
/// aca en tiempo de ejecucion.
///
/// Ver `docs/decisiones/002-gestion-de-secretos.md` para el porque.
enum AppSecrets {

  /// Claves disponibles, con el nombre exacto que tienen en el Info.plist.
  enum Key: String, CaseIterable {
    case steamAPIKey = "SteamAPIKey"
    case steamID = "SteamID"
    case itadAPIKey = "ITADAPIKey"

    /// Nombre de la variable en el archivo `.xcconfig`, para poder decirle al
    /// usuario exactamente que linea le falta.
    var xcconfigName: String {
      switch self {
      case .steamAPIKey: "STEAM_API_KEY"
      case .steamID: "STEAM_ID"
      case .itadAPIKey: "ITAD_API_KEY"
      }
    }
  }

  /// Error con un mensaje accionable, para mostrarlo en pantalla en vez de
  /// dejar que la app se caiga.
  struct MissingSecretError: LocalizedError, Equatable {
    let key: Key

    var errorDescription: String? {
      String(localized: "Falta la clave \(key.xcconfigName).", comment: "Falta un secreto de configuracion")
    }

    var recoverySuggestion: String? {
      String(
        localized: """
          Copia Config/Secrets.example.xcconfig a Config/Secrets.xcconfig y \
          escribe el valor de \(key.xcconfigName). Las instrucciones estan en el \
          README.
          """,
        comment: "Como configurar un secreto que falta"
      )
    }
  }

  /// Devuelve el valor de una clave, o lanza `MissingSecretError` si no esta.
  ///
  /// - Parameters:
  ///   - key: La clave a leer.
  ///   - bundle: Bundle de donde leer. Se puede inyectar otro en las pruebas.
  static func value(for key: Key, in bundle: Bundle = .main) throws -> String {
    guard let raw = bundle.object(forInfoDictionaryKey: key.rawValue) as? String else {
      throw MissingSecretError(key: key)
    }

    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
      throw MissingSecretError(key: key)
    }

    return trimmed
  }

  /// Claves que faltan o estan vacias. Sirve para avisar al arrancar, antes de
  /// que el usuario se tope con un error a mitad de una pantalla.
  static func missingKeys(in bundle: Bundle = .main) -> [Key] {
    Key.allCases.filter { key in
      (try? value(for: key, in: bundle)) == nil
    }
  }
}
